src/hash_pcql.c

```diff
@@ -0,0 +1,322 @@
+/*
+ * Datastructures: GAP package providing common datastructures.
+ * Licensed under the GPL 2 or later.
+ */
+
+#include "hash_pcql.h"
+
+#include "src/intfuncs.h"
+
+
+#define PERTURB_SHIFT           5
+#define LOADFACTOR_NUMERATOR    7
+#define LOADFACTOR_DENOMINATOR  10
+
+UInt PCQL_Capacity_RNam = 0;
+UInt PCQL_Deleted_RNam = 0;
+UInt PCQL_Size_RNam = 0;
+UInt PCQL_Keys_RNam = 0;
+UInt PCQL_Vals_RNam = 0;
+UInt PCQL_HashFun_RNam = 0;
+UInt PCQL_EqFun_RNam = 0;
+
+
+Obj MyHash(Obj op)
+{
+    Int n = SIZE_OBJ(op);
+
+    if (IS_INTOBJ(op))
+        return op;
+
+#ifdef SYS_IS_64_BIT
+    UInt8 hashout[2];
+    MurmurHash3_x64_128( (const void *)ADDR_OBJ(op), (int)n, 0, (void *)hashout);
+    return INTOBJ_INT(hashout[0] % (1UL << 60));
+#else
+    UInt4 hashout;
+    MurmurHash3_x86_32( (const void *)ADDR_OBJ(op), (int)n, 0, (void *)&hashout);
+    return INTOBJ_INT(hashout % (1UL << 28));
+#endif
+}
+
+static Int _PCQL_Hash_Lookup_intern(Obj ht,
+                                    Obj keys,
+                                    Obj key,
+                                    Obj hash,
+                                    Obj eqfun,
+                                    Int mask,
+                                    Int key_is_unique,
+                                    Int create)
+{
+    Obj tmp;
+    Int idx, perturb;
+    Int first_free = 0;
+
+    perturb = INT_INTOBJ(hash);
+    idx = perturb & mask;
+
+    while ((tmp = ELM_PLIST(keys, idx + 1))) {
+        if (tmp == Fail) {
+            if (first_free == 0)
+                first_free = idx + 1;
+        } else if (!key_is_unique &&
+                (eqfun ? (True == CALL_2ARGS(eqfun, tmp, key)) : EQ(tmp, key)) ) {
+            return idx + 1;
+        }
+
+        idx = (5 * idx + perturb + 1) & mask;
```

Comment:
This "second hash" should be factored out into a separeate function, so that we can more easily experiment with alternatives.